include/seqan3/alignment/data_structure/scoring_scheme_base.hpp

```diff
@@ -0,0 +1,220 @@
+    using matrix_size_t = uint8_t;
+    //!\brief Size of the matrix dimensions (i.e. size of the alphabet).
+    static constexpr matrix_size_t matrix_size = alphabet_size_v<alphabet>;
+    //!\brief Type of the internal matrix (a two-dimensional array).
+    using matrix_t = std::array<std::array<score_t, matrix_size>, matrix_size>;
+
+private:
+    //!\brief Befriend derived_t so it can instantiate.
+    friend derived_t;
+
+    //!\publicsection
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr scoring_scheme_base(scoring_scheme_base const &) noexcept = default;
+    constexpr scoring_scheme_base(scoring_scheme_base &&) noexcept = default;
+    constexpr scoring_scheme_base & operator=(scoring_scheme_base const &) noexcept = default;
+    constexpr scoring_scheme_base & operator=(scoring_scheme_base &&) noexcept = default;
+
+    //!\brief The default constructor (delegates to set_scheme_hamming()).
+    constexpr scoring_scheme_base() noexcept
+    {
+        set_scheme_hamming();
+    }
+
+    /*!\brief Constructor for the simple scheme (delegates to set_scheme_simple()).
+     * \copydetails set_scheme_simple()
+     */
+    constexpr scoring_scheme_base(match_score const ms, mismatch_score const mms) noexcept
+    {
+        set_scheme_simple(ms, mms);
+    }
+
+    /*!\brief Constructor for a custom scheme (delegates to set_scheme_custom()).
+     * \copydetails set_scheme_custom()
+     */
+    constexpr scoring_scheme_base(matrix_t const & _matrix) noexcept
+    {
+        set_scheme_custom(_matrix);
+    }
+    //!\}
+
+public:
+    /*!\name Scheme selection
+     * \{
+     */
+    //!\brief Set the hamming scheme, a variant of the simple scheme where match is scored `1` and mismatch `-1`.
+    constexpr void set_scheme_hamming() noexcept
+    {
+        set_scheme_simple(match_score{1}, mismatch_score{-1});
+    }
+
+    /*!\brief Set the simple scheme (everything is either match or mismatch).
+     * \param[in]  ms Matches shall be given this value (of type seqan3::match_score).
+     * \param[in] mms Mismatches shall be given this value (of type seqan3::mismatch_score).
+     */
+    constexpr void set_scheme_simple(match_score const ms, mismatch_score const mms)  noexcept
+    {
+        for (matrix_size_t i = 0; i < matrix_size; ++i)
+            for (matrix_size_t j = 0; j < matrix_size; ++j)
+                matrix[i][j] = (i == j) ? static_cast<score_t>(ms) : static_cast<score_t>(mms);
+    }
+
+    /*!\brief Set a custom scheme by passing a full matrix with arbitrary content.
+     * \param[in] _matrix A full matrix that is copied into the scheme.
+     */
+    constexpr void set_scheme_custom(matrix_t const & _matrix) noexcept
+    {
+        ranges::copy(_matrix, ranges::begin(matrix));
+    }
+    //!\}
+
+    /*!\name Accessors
+     * \{
+     */
+    /*!\brief Score two letters (either two nucleotids or two amino acids).
```

Comment:
> (either two nucleotids or two amino acids)

Other alphabets (modelling the alphabet concept) should be also fine or not?

Reply:
No, we only have these two and the scpecialisations inherit this documentation so I wanted to point out that it is either two of nucl or two of aa, but not one of each or random alphabets.